torch/distributed/_spmd/parallel_mode.py

```diff
@@ -0,0 +1,101 @@
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Dict, List, Tuple
+
+import torch
+import torch.distributed as dist
+import torch.utils._pytree as pytree
+from torch.distributed._spmd.distribute import _convert_to_distributed, Schema
+from torch.distributed._tensor import DeviceMesh, Placement, Replicate, Shard
+
+from torch.fx import GraphModule
+
+
+class ParallelMode(ABC):
+    """Basic Parallel Mode interface."""
+
+    @abstractmethod
+    def expand(
+        self,
+        gm: GraphModule,
+        params_and_buffers: Dict[str, Any],
+        named_states: Dict[str, Any],
+        args: Tuple[Any, ...],
+        kwargs: Dict[str, Any],
+    ) -> GraphModule:
+        """expand a single device graph to a distributed graph."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def optimize(self, gm: GraphModule) -> GraphModule:
+        """optimize a distributed graph with a set of optimization passes"""
+        # TODO: add more necessary arguments to this interface.
+        raise NotImplementedError()
+
+    def configure_optimization_passes(self, _: List[Callable]) -> None:
+        """a way to configure optimization passes per parallel mode"""
+        raise NotImplementedError()
+
+
+class DTensorFallbackMode(ParallelMode):
+    """The DTensor fallback parallel mode. It's replicating the parameters
+    and shard the inputs to represent DDP like behavior, it's currently
+    a transitent mode before we move to the new data parallel expansion.
+    """
+
+    def __init__(self):
+        self._placements_override: Dict[int, List[Placement]] = {}
+        self._optimization_passes: List[Callable] = []
+
+    def expand(
+        self,
+        gm: GraphModule,
+        params_and_buffers: Dict[str, Any],
```

Comment:
will this support hybrid paradigms? I assume yep, we basically just need to tag the params and buffers accordingly and then users can plugin a parallel mode implementation they prefer for that?

Reply:
My view on this: if user don't tag using composable API, we by default use a default mixed sharding. If user tag with composable, then in the composable we set the DataParallelMode to its corresponding style (replicate or fully shard), the expand code would expand correspondingly